Implement chunk section palettes

Pokechu22 opened the pull request:
This PR implements palettes for individual chunk sections, which should improve performance (bandwidth and memory wise), both on the server _and_ client sides.  I don't have any concrete benchmarks for this, but it should in most cases lead to chunk data being 4/13ths the size (when possible, though larger when needed).

The algorithm I'm using here is fairly simple:

1. Iterate through all blocks in the section.  For each, add the type to a FastUtil `IntList`, if the list doesn't already contain it.  This creates the palette.
2. Once the palette has been built, calculate the base-2 logarithm of the size to get the bits per block.
3. If the bits per block is less than 4, increase it to 4 (as required for the client).
4. If the bits per block is more than 8, set it to 13, and remove the palette.
5. Create a `VariableValueArray` with the bits per block.
6. Iterate through all blocks in the section, again.  For each, put it into the `VariableValueArray` the appropriate type (based off the brute-force based method `palette.indexOf(type)` if the palette isn't present) at the correct index.

Similarly, there's logic to expand the size of the palette when `set` is called, if needed.

That algorithm may be improvable (potentially by changing the order of the palette, so that something faster than `IntArrayList.indexOf` could be used), but it's a decent enough algorithm for the moment (if there are any suggestions on how to make it better, let me know!)

One thing to note is that the palette will slowly get larger within a chunk section until it is saved and reloaded; palette IDs aren't ever "freed" automatically.  This isn't ideal, but it's more expensive to try and automatically free them (and remap all of the items in the `VariableValueArray` when doing so) as far as I can tell.  Again, there may be better algorithms for that.  I did add an `optimize` method that manually rebalances the palette, but it isn't called anywhere yet (and may not need to be called anywhere).

Here's a test world that should be good for this (it contains sections with 4, 5, 6, 7, 8, and 13 bits per block, in large all-stone chunk columns): [TestWorldAllStone.zip](https://github.com/GlowstoneMC/Glowstone/files/826504/TestWorldAllStone.zip)

I made this PR to check my understanding of the palette system (while I was writing the [revised chunk format article](http://wiki.vg/User:Pokechu22/Chunk_Format)).

turt2live commented:
@PaulBGD When interacting with the noteblocks the sound duration is incorrect (not completed) and the full range of pitches is not accessible (only about 1/4 of the range is used).

turt2live commented:
@PaulBGD The last recorded activity on this pull request was 15 days ago, with an issue being identified 4 days ago. Please resolve the concern as soon as possible, thanks.

PaulBGD commented:
As you were going down the PRs, I was getting frightened as you approached mine. Anyhow, I'm trying to find fixes to both of the issues. I'm not sure why MC isn't playing the full sound duration, when I'm telling it to.

turt2live commented:
Well, the fear should have escaped you now that it's been poked :P 

As for the issues, it may be playing a shorter sound due to some weird logic. I would attempt at fixing the reduced possible range of values and hope that fixes the issue (it may be creating an unintentional rapid fire, causing sounds to sound as cut off).

turt2live commented:
@PaulBGD any update on the changes requested?

PaulBGD commented:
@turt2live I've been unable to find the cause.

mamifsidtect commented:
I know it's probably not my place to check, but how is the progression with this issue?

PaulBGD commented:
@mamifsidtect busy, if anyone would like to try to fix the changes @turt2live talked about then I'd be fine with that.

turt2live commented:
For the record, if anyone does decide to assist, please do the changes on your own fork and PR them to @PaulBGD's branch, leaving a comment here to indicate as such.

turt2live commented:
@PaulBGD Any updates on this?

Tonodus commented on src/main/java/net/glowstone/net/handler/play/player/DiggingHandler.java:
```diff
@@ -101,6 +101,9 @@ public void handle(GlowSession session, DiggingMessage message) {
             BlockPlacementHandler.revert(player, block);
+        } else {
+            BlockType blockType = ItemTable.instance().getBlock(block.getType());
+            blockType.leftClickBlock(player, block, holding);
```
Missing null check?

turt2live commented:
@PaulBGD This is a final warning: Please update your pull request as soon as possible to have review continue. Thanks!

turt2live commented:
Closing this as inactive. It's been months without communication and no visible progress. If this is not inactive, please reply as soon as possible @PaulBGD.

If anyone else would like to pick up on this please open a new PR with the appropriate changes. Thank you for your contribution!

PaulBGD commented:
This is inactive.

Pokechu22 committed:
Move chunk code into its own package, and ChunkSection into its own file

Pokechu22 committed:
Implement a chunk section palette

This is a lot of refactoring, including hiding a bunch of the fields in ChunkSection and moving writing logic there.  It also adds a few new unit tests, and allows resizing VariableValueArray (and corrects some typos there).

mastercoms commented on src/main/java/net/glowstone/chunk/GlowChunk.java:
```diff
@@ -242,7 +246,7 @@ public void initializeSections(ChunkSection... initSections) {
 
-        sections = new ChunkSection[DEPTH / SEC_DEPTH];
+        sections = new ChunkSection[SEC_COUNT];
         System.arraycopy(initSections, 0, sections, 0, Math.min(sections.length, initSections.length));
```
Do you think it would be beneficial to add a method to init an individual section by index?

Pokechu22 replied:
I'm not entirely sure what that would look like.  Do you mean replacing `initializeSections(ChunkSection... initSections)` with `initializeSection(int index, ChunkSection section)`?  There might be _some_ benefit, but I don't know where that would be used (all places where `initializeSections` are used already build an array).  It would get rid of the array with null members being passed, though.

mastercoms replied:
Yeah, I meant for it be used to skip passing null members.

Pokechu22 replied:
Alright, I'll try to do that, shouldn't be too hard.

Pokechu22 replied:
Done (correctly, this time).

mastercoms commented on src/main/java/net/glowstone/io/anvil/AnvilChunkIoService.java:
```diff
@@ -208,27 +208,29 @@ public void write(GlowChunk chunk) throws IOException {
             CompoundTag sectionTag = new CompoundTag();
```
If you add a individual init sections method, could you add methods for reading and writing individual sections?

Pokechu22 replied:
Done.

mastercoms commented:
Looks good with my initial review. Suggestions so far (about individual chunk section handling) are not essential for merging this, but they would be nice to have.

Pokechu22 committed:
Further isolate ChunkSection code

* Move loading/saving from NBT into the ChunkSection itself
* Add a separate InitializeSection method

mastercoms commented:
@Pokechu22 Seems like this PR made lighting stop working. Everything is bright during the night.

RhvsW commented:
Before (https://github.com/GlowstoneMC/Glowstone/commit/90eb1e3d2812969381342c16d89a25c1aca0edfc)

![before](https://i.hatethe.uk/fFu4eaIyhOdu.png)

After (https://github.com/GlowstoneMC/Glowstone/commit/96bd6de81c09f202e44defcab7a9b70b837a7b1d)

![after](https://i.hatethe.uk/e8jacQBpRNqu.png)
![after underground](https://i.hatethe.uk/evc0a4Vutz7D.png)

Pokechu22 commented:
Hm, I might have broken something with sky light (I did refactor the way that it's loaded).  It's probably a simple mistake, I'll look.

mastercoms commented:
Ok thanks. I didn't think to test sky light because I saw that it was just a small refactor. Oh well. Thanks for your quick response and help for this issue.

Pokechu22 commented:
Can you take a screenshot with F3 opened?  (I don't have access to a client right now; F3 would let me see what types of light are set - SL and BL)

RhvsW commented:
Build before PR: ![before](https://i.hatethe.uk/sakRUreC3fC4.png)
Latest build: ![latest](https://i.hatethe.uk/Xw4Cb5xeXxBX.png)

Pokechu22 commented:
> Light: 15 (0 sky, 15 block)

I did not expect _that_.  Hm.  Oh, I see what happened...  I'm pre-populating the blocklight array with 0xF, when I meant to populate the skylight array with 0xF.  That's easy to fix.
